### ISSUE TITLE
Testcontainers should not run multiple disk checks at the same time

### ISSUE THREAD

**Issue by jeantil** [testcontainers/testcontainers-node#43]
Hi, 

I recently opened apocas/docker-modem#112 because I kept getting seemingly innocuous stack traces in my tests logs: 
```
 console.error node_modules/jest-jasmine2/build/jasmine/Env.js:289
    Unhandled error

  console.error node_modules/jest-jasmine2/build/jasmine/Env.js:290
    Error: (HTTP code 500) server error - a disk usage operation is already running 
        at /home/jean/dev/startups/yupwego/src/yupwego/node_modules/docker-modem/lib/modem.js:257:17
        at getCause (/home/jean/dev/startups/yupwego/src/yupwego/node_modules/docker-modem/lib/modem.js:287:7)
        at Modem.Object.<anonymous>.Modem.buildPayload (/home/jean/dev/startups/yupwego/src/yupwego/node_modules/docker-modem/lib/modem.js:256:5)
        at IncomingMessage.<anonymous> (/home/jean/dev/startups/yupwego/src/yupwego/node_modules/docker-modem/lib/modem.js:232:14)
        at IncomingMessage.emit (events.js:187:15)
        at endReadableNT (_stream_readable.js:1094:12)
        at process._tickCallback (internal/process/next_tick.js:63:19)
``` 

@ehossack recently commented that he started observing these messages after upgrading from testcontainers 2.0.0 to testcontainers 2.1.0 and suggests it may be related to the changes in #42 

It seems testcontainers is attempting to run multiple disk free checks. It's a bit surprising because I can reproduce it with a single jest test which starts the container in a `beforeAll` and stops it in an `afterAll`

**Comment by cristianrgreco** [comment]
Hi @jeantil 

Thanks for raising this issue. The expected behaviour is that testcontainers should check the disk usage once, are you sure it is running it multiple times?

Could you share the test you are running which reproduces the issue? 
Could you also please provide your node and docker version?

**Comment by ehossack** [comment]
I've been thinking about this a bit more, and I don't think this should be the default behaviour of `testcontainers`.
With the eslint CLI, there's an option `--env-info` which prints information about the execution environment, which I think is the appropriate process to diagnose issues (e.g. #41, the impetus of the change in #42). Perhaps a different factory method on `DockerodeClientFactory` like `debugInstance()` and maintaining the constructor arg in the `GenericContainer`.

If the same `src/docker-client-factory.ts` is not kept in memory (jest executions, `jest#isolateModules` or `jest#resetModules`, this error can easily happen.

**Comment by cristianrgreco** [comment]
Thanks @ehossack, that's a very useful explanation. I see now that disk usage checks can conflict with other libraries, and running them across multiple jest tests can exacerbate the issue.

My first thought is to just rollback the change, and perhaps provide more info/FAQ in the readme to help users diagnose such issues. What do you think?

**Comment by ehossack** [comment]
Tough question.
I think, yes a FAQ section would be the ideal place, since it should be the first place users check on encountering difficulty. but realistically would they 😄 
The warning is a good idea, as many libraries do this, but if it has side effects like we're seeing, one has to consider the cost of the command.
I noticed no other side effects/performance hits of this, aside from a message in the logs. 

I've since lost the code that produced this error - however, if we just trap the execution, that might be sufficient?
```
      DockerodeClientFactory.instance
        .getClient()
        .info()
        .then(info => {
          log.debug(`Docker version: ${info.version}`);

          const availableGb = info.availableMb / 1000;
          if (availableGb < 2) {
            log.warn(`Docker environment has less than 2GB free disk space`);
          }
        })
       .catch(() => {}); // swallow error, as any serious exceptions will surface on container start
```

**Comment by cristianrgreco** [comment]
My understanding is as follows:

1. If other libraries perform a disk usage operation, testcontainers fails.
2. If we perform a disk usage operation, other libraries are failing. 

Catching the error would fix the first point, but we still have the second issue. 

Is my understanding correct? If it is then I think the only solution is to get rid of it altogether.

**Comment by jeantil** [comment]
@cristianrgreco 
> The expected behaviour is that testcontainers should check the disk usage once, are you sure it is running it multiple times?
> Could you share the test you are running which reproduces the issue?

I have created a sample repo demonstrating the problem at https://github.com/jeantil/documentstore
if you run a single test all is fine 
```
DEBUG="*:*" yarn jest --detectOpenHandles src/firebase/index2.test.ts                                                                                                                                     ±[●][master]
yarn run v1.17.3
$ /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/.bin/jest --detectOpenHandles src/firebase/index2.test.ts
Determining test suites to run...  snapdragon:compiler initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/compiler.js +0ms
  snapdragon:parser initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/parser.js +0ms
  snapdragon:compiler initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/compiler.js +5ms
  snapdragon:parser initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/parser.js +0ms
  ywg:store:firebase starting container +0ms
  ywg:store:firebase container started +3s
  ywg:store:firebase test starting +61ms
  console.log src/firebase/index.ts:237
    ✔ connection to firebase f459c9ce-eba3-43e1-bbc6-c8007c03b2c1 ef8a0de2-5a09-4db1-b29d-cc490e091c15

  ywg:store:firebase test completed +864ms
  ywg:store:firebase stopping container +1ms
  ywg:store:firebase container stopped +10s
 PASS  src/firebase/index2.test.ts (15.065s)
  FirebaseStore User repository
    ✓ should delete tokens  (14385ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        15.09s, estimated 19s
Ran all test suites matching /src\/firebase\/index2.test.ts/i.
Done in 17.60s.
``` 
if you try to run all the tests one of them fails because the generic container startup fails with an exection due to the already running disk operation. note that the docker container itself does start and has to be cleaned up manually. 

```
DEBUG="*:*" yarn jest                                                                                                                                                                                     ±[●][master]
yarn run v1.17.3
$ /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/.bin/jest
Determining test suites to run...  snapdragon:compiler initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/compiler.js +0ms
  snapdragon:parser initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/parser.js +0ms
  snapdragon:compiler initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/compiler.js +5ms
  snapdragon:parser initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/parser.js +0ms
 PASS  src/FirebaseConfig.test.ts
  FirebaseConfig
    ✓ should extract all config information from env (5ms)
    ✓ should build emulator config (1ms)
    ✓ should fail on missing FIREBASE_DB_URL (2ms)
    ✓ should fail on null FIREBASE_DB_URL
    ✓ should fail on empty FIREBASE_DB_URL (1ms)
    ✓ should fail on missing FIREBASE_PROJECT_ID (1ms)
    ✓ should fail on null FIREBASE_PROJECT_ID
    ✓ should fail on empty FIREBASE_PROJECT_ID
    ✓ should fail on missing FIREBASE_CLIENT_EMAIL
    ✓ should fail on null FIREBASE_CLIENT_EMAIL
    ✓ should fail on empty FIREBASE_CLIENT_EMAIL
    ✓ should fail on missing FIREBASE_PRIVATE_KEY (1ms)
    ✓ should fail on null FIREBASE_PRIVATE_KEY
    ✓ should fail on empty FIREBASE_PRIVATE_KEY

2019-12-07T15:39:58.549Z ywg:store:firebase starting container
2019-12-07T15:39:58.587Z ywg:store:firebase starting container
2019-12-07T15:39:58.616Z ywg:store:firebase test starting
2019-12-07T15:39:58.617Z ywg:store:firebase stopping container
Sat, 07 Dec 2019 15:39:58 GMT snapdragon:compiler initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/compiler.js
Sat, 07 Dec 2019 15:39:58 GMT snapdragon:parser initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/parser.js
Sat, 07 Dec 2019 15:39:58 GMT snapdragon:compiler initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/compiler.js
Sat, 07 Dec 2019 15:39:58 GMT snapdragon:parser initializing /home/jean/dev/startups/yupwego/src/gitlabci/node_modules/snapdragon/lib/parser.js
 FAIL  src/firebase/index2.test.ts
  FirebaseStore User repository
    ✕ should delete tokens  (33ms)

  ● FirebaseStore User repository › should delete tokens 

    (HTTP code 500) server error - a disk usage operation is already running

      at node_modules/docker-modem/lib/modem.js:257:17
      at getCause (node_modules/docker-modem/lib/modem.js:287:7)
      at Modem.Object.<anonymous>.Modem.buildPayload (node_modules/docker-modem/lib/modem.js:256:5)
      at IncomingMessage.<anonymous> (node_modules/docker-modem/lib/modem.js:232:14)

  ● FirebaseStore User repository › should delete tokens 

    TypeError: Cannot read property 'tokens' of undefined

      38 |       expires: 0
      39 |     };
    > 40 |     await store.tokens.save(token.id, token);
         |                 ^
      41 |     const saved = await store.tokens.get(token.id);
      42 |     await store.tokens.delete(token.id);
      43 |     const actual = await store.tokens.getOpt(token.id);

      at src/firebase/index2.test.ts:40:17
      at step (src/firebase/index2.test.ts:33:23)
      at Object.next (src/firebase/index2.test.ts:14:53)
      at src/firebase/index2.test.ts:8:71
      at Object.<anonymous>.__awaiter (src/firebase/index2.test.ts:4:12)
      at Object.<anonymous> (src/firebase/index2.test.ts:32:31)

  ● FirebaseStore User repository › should delete tokens 

    TypeError: Cannot read property 'close' of undefined

      26 |     //stop docker container
      27 |     debug('stopping container');
    > 28 |     await store.close();
         |                 ^
      29 |     await container.stop();
      30 |     debug('container stopped');
      31 |   });

      at src/firebase/index2.test.ts:28:17
      at step (src/firebase/index2.test.ts:33:23)
      at Object.next (src/firebase/index2.test.ts:14:53)
      at src/firebase/index2.test.ts:8:71
      at Object.<anonymous>.__awaiter (src/firebase/index2.test.ts:4:12)
      at Object.<anonymous> (src/firebase/index2.test.ts:25:13)

2019-12-07T15:40:02.912Z ywg:store:firebase container started

ReferenceError: You are trying to `import` a file after the Jest environment has been torn down.



      at node_modules/firebase-admin/lib/firebase-app.js:260:36
      at FirebaseApp.Object.<anonymous>.FirebaseApp.ensureService_ (node_modules/firebase-admin/lib/firebase-app.js:351:23)
      at FirebaseApp.Object.<anonymous>.FirebaseApp.firestore (node_modules/firebase-admin/lib/firebase-app.js:259:28)
      at new FirebaseStore (src/firebase/index.ts:310:27)
      at src/firebase/index2.test.ts:23:13
(node:20855) UnhandledPromiseRejectionWarning: TypeError: Caught error after test environment was torn down

firestoreService is not a constructor
(node:20855) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 5)
(node:20855) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
2019-12-07T15:40:02.937Z ywg:store:firebase container started
2019-12-07T15:40:03.014Z ywg:store:firebase test starting
  console.log src/firebase/index.ts:237
    ✔ connection to firebase 5592925f-f079-45a7-91f6-640220117fe4 2e203c28-7d9d-4fa0-b14d-9954b71c0625

2019-12-07T15:40:04.030Z ywg:store:firebase test completed
2019-12-07T15:40:04.030Z ywg:store:firebase stopping container
2019-12-07T15:40:14.304Z ywg:store:firebase container stopped
 PASS  src/firebase/index.test.ts (16.492s)
  FirebaseStore User repository
    ✓ should store tokens  (15756ms)

Test Suites: 1 failed, 2 passed, 3 total
Tests:       1 failed, 15 passed, 16 total
Snapshots:   12 passed, 12 total
Time:        16.713s, estimated 17s
Ran all test suites.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```  

Jest runs different test suites in parallel to speed up overall test execution time, this leads to 2 generic containers starting at the same time and thus two calls to check disk space. This should not make tests fail.

> Could you also please provide your node and docker version?
```
>>> node --version                                                                                                                                                                                            
v10.14.2
>>> docker version                                                                                                                                                                                            
Client:
 Version:           19.03.4-ce
 API version:       1.40
 Go version:        go1.13.1
 Git commit:        9013bf583a
 Built:             Sat Oct 19 04:40:07 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.4-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.1
  Git commit:       9013bf583a
  Built:            Sat Oct 19 04:39:38 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.3.0.m
  GitCommit:        d50db0a42053864a270f648048f9a8b4f24eced3.m
 runc:
  Version:          1.0.0-rc9
  GitCommit:        d736ef14f0288d6993a1845745d6756cfc9ddd5a
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

**Comment by cristianrgreco** [comment]
Hi @jeantil 

I have released 2.2.0 which reverts the disk check. 
I've added a "Common Issues" to the top of the README instead.

Please let me know if this has resolved the issue for you.

**Comment by jeantil** [comment]
Hi @cristianrgreco 

Thanks for the new release, I have updated my sample repository and it does indeed fix the issues I reported above.
